### PR TITLE
Add Polars implementation of leaf isomorphism test

### PR DESCRIPTION
**Pull request by mmore500**
## Summary
This PR adds a Polars-based implementation of the phylogenetic leaf isomorphism test, complementing the existing Pandas-based asexual implementation. The new implementation provides the same functionality with support for both eager DataFrames and lazy evaluation.

## Key Changes
- **New module**: `phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py`
  - Implements `alifestd_test_leaves_isomorphic_polars()` function that tests if phylogenetic relationships between leaf nodes are topologically isomorphic between two phylogenies
  - Supports both `pl.DataFrame` and `pl.LazyFrame` inputs
  - Includes CLI entrypoint for comparing phylogeny files (CSV, Parquet, Feather formats)
  - Uses numba JIT compilation for performance-critical tree traversal logic

- **Comprehensive test suite**: `tests/test_phyloframe/test_legacy/test_alifestd_test_leaves_isomorphic_polars.py`
  - 294 lines of tests covering empty trees, mutation safety, positive/negative cases
  - Fuzzing tests against real phylogeny datasets and synthetic trees
  - Cross-validation with the existing Pandas asexual implementation
  - Tests for edge cases (disjoint leaves, different node counts, etc.)

- **CLI tests**: `tests/test_phyloframe/test_legacy/test_alifestd_test_leaves_isomorphic_polars_cli.py`
  - Tests for help/version output
  - Self-comparison tests (should return exit code 0)
  - Different phylogeny tests (should return exit code 1)

- **Module exports**: Updated `phyloframe/legacy/__init__.pyi` and `phyloframe/__main__.py` to expose the new function and CLI

## Implementation Details
- The algorithm works by:
  1. Converting input DataFrames to a working format (topologically sorted with contiguous IDs)
  2. Collapsing unifurcations in both trees
  3. Marking leaf nodes
  4. Building an ID mapping from matching leaf taxon labels
  5. Walking backward through the tree structure to verify ancestor mappings are consistent
- Uses numba JIT compilation (`@jit(nopython=True)`) for the tree traversal loop to ensure performance
- Handles both integer and string taxon labels
- Supports special case where `taxon_label == "id"` uses the ID column directly

https://claude.ai/code/session_01TweTScZz7Jyd9hkzPztc4a

### CODE REVIEW

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+    if isinstance(df1, pl.LazyFrame):
+        df1 = df1.collect()
+    if isinstance(df2, pl.LazyFrame):
+        df2 = df2.collect()
+
+    logging.info(
+        "- alifestd_test_leaves_isomorphic_polars: "
+        "preparing working format...",
+    )
+    df1 = _to_working_format_polars(df1)
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: just raise NotImplementedError if not topologically sorted/contiguous

Comment (mmore500) [reply phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: Look for other patterns of this in library

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+    if not alifestd_has_contiguous_ids_polars(phylogeny_df):
+        phylogeny_df = alifestd_assign_contiguous_ids_polars(phylogeny_df)
+    if not alifestd_is_topologically_sorted_polars(phylogeny_df):
+        phylogeny_df = alifestd_topological_sort_polars(phylogeny_df)
+        phylogeny_df = alifestd_assign_contiguous_ids_polars(phylogeny_df)
+    return phylogeny_df
+
+
+def alifestd_test_leaves_isomorphic_polars(
+    df1: typing.Union[pl.DataFrame, pl.LazyFrame],
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: Just type hint DataFrame; look for existing pattern for this

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+) -> bool:
+    """Test if phylogenetic relationships between leaf nodes are topologically
+    isomorphic between two phylogenies.
+
+    See Also
+    --------
+    alifestd_test_leaves_isomorphic_asexual :
+        Pandas-based implementation.
+    """
+    if isinstance(df1, pl.LazyFrame):
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: Do NOT collect the entire dataframe; if individual columns are truly needed collect only those and as late as possible; extensively review other polars code in the library for guidance on appropriate patterns!

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    """Walk ids backward and propagate the leaf-induced id mapping toward the
+    roots, returning False if any node's ancestor mapping is inconsistent.
+
+    Both phylogenies are assumed to be topologically sorted with contiguous
+    ids. ``id_map`` should be initialized so that ``id_map[id1] = id2`` for
+    every leaf in df1 with matching taxon label in df2; ``id_map_set`` is a
+    parallel boolean mask indicating which entries of ``id_map`` are valid.
+    """
+    n = jit_numpy_int64_t(len(ancestor_ids1))
+    for offset in range(n):
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: Follow existing pattern in the library for reversed enumerate

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+    if len(ancestor_ids1) != len(ancestor_ids2):
+        return False
+
+    n = len(ancestor_ids1)
+    id_map = np.zeros(n, dtype=np.int64)
+    id_map_set = np.zeros(n, dtype=np.bool_)
+    leaf_ids1 = leaf_pairs["id"].to_numpy()
+    leaf_ids2 = leaf_pairs["id2"].to_numpy()
+    id_map[leaf_ids1] = leaf_ids2
+    id_map_set[leaf_ids1] = True
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: This can be computed using array op inside jit function

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    logging.info(
+        "- alifestd_test_leaves_isomorphic_polars: marking leaves...",
+    )
+    df1 = alifestd_mark_leaves_polars(df1)
+    df2 = alifestd_mark_leaves_polars(df2)
+
+    if taxon_label == "id":
+        df1 = df1.with_columns(taxon_label=pl.col("id"))
+        df2 = df2.with_columns(taxon_label=pl.col("id"))
+        taxon_label = "taxon_label"
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: Need to use aliases that won’t conflict with user data

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+
+    Both phylogenies are assumed to be topologically sorted with contiguous
+    ids. ``id_map`` should be initialized so that ``id_map[id1] = id2`` for
+    every leaf in df1 with matching taxon label in df2; ``id_map_set`` is a
+    parallel boolean mask indicating which entries of ``id_map`` are valid.
+    """
+    # iterate over ids from back to front
+    for id1_r, ancestor_id1 in enumerate(ancestor_ids1[::-1]):
+        id1 = jit_numpy_int64_t(len(ancestor_ids1) - 1 - id1_r)
+        if not id_map_set[id1]:
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: Do this check first, ideally using array operations, possibly outside the jit loop

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+    id_map[leaf_ids1] = leaf_ids2
+    id_map_set[leaf_ids1] = True
+
+    # All leaves were mapped above; every inner node will have its mapping
+    # propagated up by some descendant before we process it (after
+    # collapse_unifurcations there are no leaf-less subtrees), so the walk
+    # below need not re-verify ``id_map_set[id1]`` per iteration.
+
+    # iterate over ids from back to front
+    for id1_r, ancestor_id1 in enumerate(ancestor_ids1[::-1]):
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: Just do range with -1’s

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+    # propagated up by some descendant before we process it (after
+    # collapse_unifurcations there are no leaf-less subtrees), so the walk
+    # below need not re-verify ``id_map_set[id1]`` per iteration.
+
+    # iterate over ids from back to front
+    for id1 in range(n - 1, -1, -1):
+        ancestor_id1 = ancestor_ids1[id1]
+        id2 = id_map[id1]
+        ancestor_id2 = ancestor_ids2[id2]
+        if id_map_set[ancestor_id1]:
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: If not then else if

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+    # collapse_unifurcations leaves gapped ids; recompact so the jit walk
+    # below can index by id.
+    df1 = alifestd_assign_contiguous_ids_polars(df1).lazy()
+    df2 = alifestd_assign_contiguous_ids_polars(df2).lazy()
+
+    logging.info(
+        "- alifestd_test_leaves_isomorphic_polars: marking leaves...",
+    )
+    # use namespaced internal column names to avoid clobbering user data
+    is_leaf_col = "_phyloframe_test_leaves_isomorphic_polars_is_leaf"
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: You can use default is leaf; use the pattern for testing if it’s present and regenerating it if need be

Define label Col before using

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+        .collect()
+    )
+
+    if leaves1[label_col].n_unique() != leaves1.height:
+        raise ValueError("taxon labels in df1 must be unique among leaves")
+    if leaves2[label_col].n_unique() != leaves2.height:
+        raise ValueError("taxon labels in df2 must be unique among leaves")
+
+    leaf_pairs = leaves1.join(leaves2, on=label_col, how="inner")
+    if leaf_pairs.height != leaves1.height or leaves1.height != leaves2.height:
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: Use count leaves polars

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+    leaves1 = (
+        df1.filter(pl.col("is_leaf"))
+        .select([pl.col("id"), pl.col(taxon_label).alias(label_col)])
+        .collect()
+    )
+    leaves2 = (
+        df2.filter(pl.col("is_leaf"))
+        .select(
+            [
+                pl.col("id").alias("id2"),
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: id2 may conflict

Comment (mmore500) [reply phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: Also why is this necessary? To have the id2 col

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+    """
+    df1 = df1.lazy()
+    df2 = df2.lazy()
+
+    if df1.limit(1).collect().is_empty() and df2.limit(1).collect().is_empty():
+        return True
+
+    df1 = alifestd_try_add_ancestor_id_col_polars(df1)
+    df2 = alifestd_try_add_ancestor_id_col_polars(df2)
+
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: raise NotImplementedError if ancestor id Col not present

Comment (mmore500) [reply phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: After doing try add ancestor id col

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+        raise NotImplementedError("topological sort not yet supported")
+
+    if not alifestd_has_contiguous_ids_polars(
+        df1
+    ) or not alifestd_has_contiguous_ids_polars(df2):
+        raise NotImplementedError("non-contiguous ids not yet supported")
+
+    # collapse_unifurcations_polars and assign_contiguous_ids_polars reject
+    # the ancestor_list column; drop it inline (no-op if absent).
+    df1 = df1.select(pl.exclude("ancestor_list"))
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: Use pipe pattern

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+) -> bool:
+    """Test if phylogenetic relationships between leaf nodes are topologically
+    isomorphic between two phylogenies.
+
+    See Also
+    --------
+    alifestd_test_leaves_isomorphic_asexual :
+        Pandas-based implementation.
+    """
+    df1 = df1.lazy()
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: Just do df.lazy().call()’s

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+    df2 = df2.select(pl.exclude("ancestor_list"))
+
+    logging.info(
+        "- alifestd_test_leaves_isomorphic_polars: "
+        "collapsing unifurcations...",
+    )
+    df1 = alifestd_collapse_unifurcations_polars(
+        df1,
+        ignore_topological_sensitivity=True,
+        drop_topological_sensitivity=False,
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: Only need drop topologically sensitivity I think

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+
+
+def _canonicalize(df: pl.LazyFrame) -> pl.LazyFrame:
+    """Validate preconditions and prepare ``df`` for the leaf-isomorphism walk.
+
+    Adds ``ancestor_id`` (if absent and derivable from ``ancestor_list``),
+    drops ``ancestor_list``, collapses unifurcations, recompacts ids so
+    the jit walk can index by id, and ensures ``is_leaf`` is present.
+    """
+    df = alifestd_try_add_ancestor_id_col_polars(df)
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: Do inline

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+
+    logging.info(
+        "- alifestd_test_leaves_isomorphic_polars: collecting leaf ids...",
+    )
+    # sorting both leaf sets by taxon label aligns matching leaves
+    # element-wise, side-stepping a join (and the column-aliasing it would
+    # require to disambiguate "id" between the two frames).
+    leaf_cols = ["id"] if taxon_label == "id" else ["id", taxon_label]
+    leaves1_sorted = (
+        df1.filter(pl.col("is_leaf"))
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: do lazy cast here

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+    if n_leaves1 != n_leaves2:
+        return False
+
+    logging.info(
+        "- alifestd_test_leaves_isomorphic_polars: collecting leaf ids...",
+    )
+    # sorting both leaf sets by taxon label aligns matching leaves
+    # element-wise, side-stepping a join (and the column-aliasing it would
+    # require to disambiguate "id" between the two frames).
+    leaf_cols = ["id"] if taxon_label == "id" else ["id", taxon_label]
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: use set

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 20]
```diff
@@ -0,0 +1,278 @@
...
+    alifestd_test_leaves_isomorphic_asexual :
+        Pandas-based implementation.
+    """
+    if (
+        df1.lazy().limit(1).collect().is_empty()
+        and df2.lazy().limit(1).collect().is_empty()
+    ):
+        return True
+
+    df1 = alifestd_try_add_ancestor_id_col_polars(df1.lazy())
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: Don’t need to cast lazy here

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 21]
```diff
@@ -0,0 +1,283 @@
...
+    )
+    leaves2_sorted = (
+        df2.filter(pl.col("is_leaf"))
+        .sort(taxon_label)
+        .select(leaf_cols)
+        .collect()
+    )
+
+    if leaves1_sorted[taxon_label].n_unique() != n_leaves1:
+        raise ValueError("taxon labels in df1 must be unique among leaves")
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]:
```suggestion
        raise ValueError("taxon labels must be unique among leaves")
```

File: phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py [hunk 22]
```diff
@@ -0,0 +1,283 @@
...
+    # collapse_unifurcations_polars and assign_contiguous_ids_polars reject
+    # the ancestor_list column; drop it inline (no-op if absent).
+    df1 = (
+        df1.select(pl.exclude("ancestor_list"))
+        .pipe(
+            alifestd_collapse_unifurcations_polars,
+            drop_topological_sensitivity=True,
+        )
+        .pipe(alifestd_assign_contiguous_ids_polars)
+        .lazy()
```
Comment (mmore500) [inline phyloframe/legacy/_alifestd_test_leaves_isomorphic_polars.py]: NO!! do lazy cast the opening of statements that collect

**Review by mmore500 — COMMENTED** [review]
lgtm, ready to merge after Ci
